File: medium/209_MinimumSizeSubarraySum.java
/**
 * 209. Minimum Size Subarray Sum
 * Medium
 * https://leetcode.com/problems/minimum-size-subarray-sum
 */
class Solution {
    public int minSubArrayLen(int target, int[] nums) {
        int l = nums.length;
        int left = 0;
        int right = 0;
        int sum = nums[0];
        int res = l + 1;
        while (right < l) {
            if (sum >= target) {
                if (left == right) {
                    return 1;
                } else {
                    res = Math.min(res, right - left + 1);
                    sum -= nums[left++];
                }
            } else if (right == l - 1) {
                break;
            } else {
                sum += nums[++right];
                if (right - left + 1 > res) {
                    sum -= nums[left++];
                }
            }
        }

        return res <= l ? res : 0;
    }
}

File: medium/209_MinimumSizeSubarraySum.py
# 209. Minimum Size Subarray Sum
# medium
# https://leetcode.com/problems/minimum-size-subarray-sum

class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        n = len(nums)
        l = 0
        summ = 0
        minimal_length = float('inf')
        for r in range(n):
            summ += nums[r]

            while summ >= target:
                minimal_length = min(minimal_length, r - l + 1)
                if minimal_length == 1:
                    return 1
                summ -= nums[l]
                l += 1
        return minimal_length if minimal_length != float('inf') else 0
